src/qs8-vadd/gen/qs8-vadd-minmax-hvx-u128.c

```diff
@@ -0,0 +1,154 @@
+// Auto-generated file. Do not edit!
+//   Template: src/qs8-vadd/hvx.c.in
+//   Generator: tools/xngen
+//
+// Copyright 2024 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <assert.h>
+
+#include <hvx_hexagon_protos.h>
+#include <hexagon_protos.h>
+#include <hexagon_types.h>
+
+#include <xnnpack/vbinary.h>
+#include <xnnpack/intrinsics-polyfill.h>
+
+void xnn_qs8_vadd_minmax_ukernel__hvx_u128(
+    size_t batch,
+    const int8_t* input_a,
+    const int8_t* input_b,
+    int8_t* output,
+    const union xnn_qs8_add_minmax_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(batch != 0);
+  assert(batch % sizeof(int8_t) == 0);
+  assert(input_a != NULL);
+  assert(input_b != NULL);
+  assert(output != NULL);
+
+  const HVX_Vector vbias = Q6_V_vsplat_R(*((int32_t *) &params->hvx.bias));
+  const HVX_Vector va_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.a_multiplier));
+  const HVX_Vector va_multiplier_lo = Q6_V_vsplat_R(*((uint16_t *) &params->hvx.a_multiplier_lo));
+  const HVX_Vector va_multiplier_hi = Q6_V_vsplat_R(*((uint16_t *) &params->hvx.a_multiplier_hi));
+  const HVX_Vector vb_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.b_multiplier));
+  const HVX_Vector vb_multiplier_lo = Q6_V_vsplat_R(*((uint16_t *) &params->hvx.b_multiplier_lo));
+  const HVX_Vector vb_multiplier_hi = Q6_V_vsplat_R(*((uint16_t *) &params->hvx.b_multiplier_hi));
+  const int32_t first_shift = params->hvx.first_shift;
+  const int32_t rest_shift = params->hvx.rest_shift;
+  const HVX_Vector voutput_zero_point = Q6_Vh_vsplat_R(*((int16_t *) &params->hvx.output_zero_point));
+  const HVX_Vector voutput_min = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_min));
+  const HVX_Vector voutput_max = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_max));
+  int8_t* ptr_o = output;
+
+  for (; batch >= 128 * sizeof(int8_t); batch -= 128 * sizeof(int8_t)) {
+    HVX_Vector va0 = *((HVX_UVector*)input_a);
+    HVX_Vector vb0 = *((HVX_UVector*)input_b);
+    input_a += 128;
+    input_b += 128;
+
+    // widen: 8-bit to 16-bit
+    HVX_VectorPair va0_i16 = Q6_Wh_vunpack_Vb(va0);
+    HVX_VectorPair vb0_i16 = Q6_Wh_vunpack_Vb(vb0);
+    HVX_Vector va0_lo = Q6_V_lo_W(va0_i16);
+    HVX_Vector va0_hi = Q6_V_hi_W(va0_i16);
+    HVX_Vector vb0_lo = Q6_V_lo_W(vb0_i16);
+    HVX_Vector vb0_hi = Q6_V_hi_W(vb0_i16);
+
+    // vacc = va * va_multiplier + vb * vb_multiplier with widening to 32-bit
+    HVX_Vector va0_lo_mul_even = Q6_V_vzero();
```

Comment:
Can you just use the non-accumulating vmpyie to start with, rather than accumulating with zero?

Reply:
This was good idea as I needed 'left shift by 16bits' after multiplying high elements. Thank you for the comments!

I used `vmpyieo(Vu, VV)` to multiply even elements of Vu by odd elements of Vv, then shift the result left by 16 bits. Unfortunately, I could not avoid using vshuffe as this intrinsic picks up odd elements from VV but multiplier_hi were even elements. 

```
static XNN_INTRINSIC
HVX_Vector Q6_Vw_vmpyie_VwVh(HVX_Vector multiplier_lo, HVX_Vector multiplier_hi, HVX_Vector vin)
{
    multiplier_hi = Q6_Vh_vshuffe_VhVh(multiplier_hi, multiplier_hi);
    HVX_Vector vout = Q6_Vw_vmpyieo_VhVh(vin, multiplier_hi);
    vout = Q6_Vw_vmpyieacc_VwVwVh(vout, multiplier_lo, vin);

    return vout;
}
```

Reply:
It still seems like there is probably a way to avoid one of the two adds here (or maybe even both if you initialize with the bias) but I'm not sure and I think maybe it'd be better to merge this and work on this in a fine-tuning PR later (and it's low priority vs. working on other kernels).

Reply:
I quickly tried and seems working and I could remove one add instruction + va_multiplier_hi/lo as well! I initialized even_mul with bias then use `vmypieacc`, but odd_mul remained same. Changes look like this:

```
HVX_Vector va0_hi_mul_even = vbias;
va0_hi_mul_even = Q6_Vw_vmpyieacc_VwVwVh(va0_hi_mul_even, va_multiplier, va0_hi);
HVX_Vector va0_hi_mul_odd = Q6_Vw_vadd_VwVw(vbias, Q6_Vw_vmpyio_VwVh(va_multiplier, va0_hi));
```

What do you think?